Fixed rules for booking

Daniel0K opened the pull request:
Добавил крайние значения и переписал логику немного метода на проверку правил.

Daniel0K committed:
Добавлены крайние условия для проверки бронирования

Daniel0K committed:
quickfix tests

Daniel0K committed:
Изменил правила бронирования.
Добавил в html на календарь проверку на min

Daniel0K committed:
Tests for booking have been moved to separate file.
Added initialization startDate field for datePicker.

YaroslavTrubanov commented on src/app/pins-form/pins-form.component.html:
```diff
@@ -1,18 +1,25 @@
+  <input
+    type="date"
+    name="End"
+    min="{{ this.form.value.startDate }}"
```
тут не должно быть обращение через this

YaroslavTrubanov commented on src/app/pins-form/pins-form.component.html:
```diff
@@ -1,18 +1,25 @@
+    min="{{ this.form.value.startDate }}"
+    formControlName="endDate"
+  />
+  <button class="btn" id="#btn1" type="submit" [disabled]="form.invalid">
```
надо нормальный id

YaroslavTrubanov commented on src/app/services/booking.service.spec.ts:
```diff
@@ -0,0 +1,55 @@
+    const prevBookings = [];
+    const tempBook1 = {} as Bookings;
+    const tempBook2 = {} as Bookings;
+    newBooking.startDate = new Date(2021, 5, 21);
```
зачем разделил инициализацию и простановку значений свойствам?

YaroslavTrubanov commented on src/app/services/booking.service.spec.ts:
```diff
@@ -0,0 +1,55 @@
+    bookingService = new BookingService(mainService);
+  });
+
+  it('should check bookings date for crossings. It will be return true because no crossing', () => {
```
```suggestion
  it('should return true when there are no crossing', () => {
```

YaroslavTrubanov commented on src/app/services/booking.service.spec.ts:
```diff
@@ -0,0 +1,55 @@
+    ).toBeTruthy();
+  });
+
+  it('should check bookings date for crossings. It will be return false because crossing exist', () => {
```
```suggestion
  it('should return false when crossing exists', () => {
```

YaroslavTrubanov commented on src/app/services/booking.service.spec.ts:
```diff
@@ -0,0 +1,55 @@
+    ).toBeFalsy();
+  });
+
+  it('check if endDate less than startDate', () => {});
```
хороший тест

Daniel0K committed:
Fixed initialization at booking tests.
Fixed rule for booking crossing.
Fixed date picker at calendar.
Removed id field from Pins-form html

YaroslavTrubanov commented on src/app/services/booking.service.spec.ts:
```diff
@@ -0,0 +1,59 @@
+      endDate: new Date(2021, 5, 28),
+    };
+
+    prevBookings.push(tempBook1);
```
да сразу и засунь туда элементы, зачем так разделять

YaroslavTrubanov replied:
ты просто закрыл коммент и не исправил, элементы все также отдельно добавляются в массив, хотя можно объединить и не писать портянку кода

YaroslavTrubanov commented on src/app/pins-form/pins-form.component.ts:
```diff
@@ -25,6 +26,8 @@ export class PinsFormComponent {
       startDate: new FormControl('', [Validators.required]),
       endDate: new FormControl('', [Validators.required]),
     });
+    this.form.value.startDate = new Date().toISOString().split('T')[0];
```
че за жесть

Daniel0K committed:
Fixed initialization at booking tests fields
Booking form will be reset after submit

Daniel0K committed:
Fixed Booking rules

Daniel0K committed:
Fixed Booking rules.
Added observables.

YaroslavTrubanov commented on src/app/pins-form/pins-form.component.ts:
```diff
@@ -25,6 +27,12 @@ export class PinsFormComponent {
     });
+    this.form.value.startDate = new Date().toISOString().split('T')[0];
+    this.startDate = this.form.value.startDate;
+    this.form.get('startDate').valueChanges.subscribe((selectedValue) => {
```
вот тут ты подписался на обзервабл  valueChanges, от таких вещей еще надо отписываться в onDestroy, чтобы лишняя память не капала
но это просто коммент на будущее

Daniel0K committed:
Added inionDestroy for observable.
Changed array initialization.

YaroslavTrubanov commented on src/app/services/booking.service.spec.ts:
```diff
@@ -0,0 +1,59 @@
+      startDate: new Date(2021, 5, 21),
+      endDate: new Date(2021, 5, 22),
+    };
+    const prevBookings = [];
+    const tempBook1 = {
+      startDate: new Date(2021, 5, 25),
+      endDate: new Date(2021, 5, 26),
+    };
+    const tempBook2 = {
+      startDate: new Date(2021, 5, 27),
+      endDate: new Date(2021, 5, 28),
+    };
```
```suggestion
    const prevBookings = [
    {
      startDate: new Date(2021, 5, 25),
      endDate: new Date(2021, 5, 26),
    }, 
    {
      startDate: new Date(2021, 5, 27),
      endDate: new Date(2021, 5, 28),
    }];
```

Daniel0K committed:
Fixed 2nd booking test initialization
